### ISSUE TITLE
Prop Mutability

### ISSUE THREAD

**Issue by aparnapatil01** [ionic-team/stencil#1927]
As per documentation, https://stenciljs.com/docs/properties#prop-mutability - `It's important to know, that a Prop is by default immutable from inside the component logic. Once a value is set by a user, the component cannot update it internally.`

However, I can set new value to the `prop` without declaring it as `mutable`. 

Code

```
// index.html
<my-component color="pink"></my-component>

//my-component.tsx
import { Component, h, Prop } from '@stencil/core';

@Component({
  tag: 'my-component',
  styleUrl: 'my-component.css',
  shadow: true
})
export class MyComponent {
  @Prop() color: string; 

  componentDidLoad() {
    this.color = "New Green";
    console.log(this.color); // Prints "New Green"
  }
  render() {
    return (
      <div>{this.color}</div>
    )
  }
}
```
The above code log "New Green" to console as well render updated value of property `color`.

It would be great if you could clarify the exact purpose of `mutable` property?

**Comment by jsanz1209** [comment]
The same thing keeps happening to me.

**Comment by ralscha** [comment]
I'm also a bit confused using version 1.12.6. 
It does not matter if I omit `mutable`, or set it to `true` or to `false`. The effect is always the same. The property is mutable from inside the component. Tested with development and production builds.

**Comment by kskeerthi** [comment]
This had happened with me as well.... Is there anything that had been changed from the new version but it is not been updated in the documentation???  I had put the prop value on didload but without me specifying mutable: true, it is still updated. If someone is aware of the reason kindly let me know!!!


import { Component, ComponentInterface, h, Prop} from '@stencil/core';

@Component({
  tag: 'app-button',
  styleUrl: 'app-button.css'
  // shadow: true,
})
export class AppButton implements ComponentInterface {
@Prop() message : string = "Hello Everyone!!"
componentDidLoad(){
  
  this.message ="kkkk"
  console.log(this.message)
}
render() {
    return (
      <p>{this.message}</p>
    );
    }

}

**Comment by Sandhyah23** [comment]
Stencil [document ](https://stenciljs.com/docs/properties#prop-mutability) is confusing, prop values are mutable inside component irrespective of the 'mutable' option.

**Comment by RenanRossiDias** [comment]
I'm using 2.0.1 and experiencing the very same problem here.
Beyond that, the @Prop isn't mutating on it's parent declaration nor reflecting changed values there.

**Comment by claviska** [comment]
See my comment [here](https://github.com/ionic-team/stencil/issues/2433#issuecomment-740025554). I'm looking into why the warning isn't emitted.

> I don't believe Stencil can actually enforce immutability, but you should see a warning in dev mode when the immutable prop is changed. This doesn't appear to be happening in 2.3.0.

**Comment by RenanRossiDias** [comment]
> See my comment [here](https://github.com/ionic-team/stencil/issues/2433#issuecomment-740025554). I'm looking into why the warning isn't emitted.
> 
> > I don't believe Stencil can actually enforce immutability, but you should see a warning in dev mode when the immutable prop is changed. This doesn't appear to be happening in 2.3.0.

@claviska That should be enough to help with this issue.
Though i'm still confused about **mutable**, should it reflect it's changes to it's direct parent, acting like a two-way prop binding, or mutable only serves the purpose of 'locking' changes to a property not to be modified inside components?

**Comment by claviska** [comment]
> or mutable only serves the purpose of 'locking' changes to a property not to be modified inside components?

Any prop can be change outside of the component, immutable or not, otherwise it would be a constant.

**Comment by claviska** [comment]
While we don't stop props from mutating, we can (and always intended to) emit a warning in dev. That logic has been fixed in #2779.
